packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,181 @@
+  let vatCloseError;
+  const { vat, stderrP, exitedPKit } = worker;
+
+  try {
+    // this will error out since we are intentionally
+    // breaking communications with the worker
+    await vat.issueStringCommand('0');
+  } catch (err) {
+    issueCommandError = err;
+  }
+
+  // wait for the worker to exit
+  const [strErr, { code: vatExitCode, signal: vatExitSignal }] =
+    await Promise.all([stderrP, exitedPKit.promise]);
+
+  try {
+    // this will error out since vat has died.
+    await vat.close();
+  } catch (err) {
+    vatCloseError = err;
+  }
+  return {
+    issueCommandError,
+    vatCloseError,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  };
+}
+
+test('xsnap-worker complains while trying to READ an answer when pipes are closed', async t => {
+  const worker = await spawnReflectiveWorker(
+    async function handleCommand(_message) {
+      // the worker is blocked on read here, so we should close "fromXsnap" pipe first
+      worker.fromXsnap.end();
+      worker.fromXsnap.destroy();
+      worker.toXsnap.end();
+      worker.toXsnap.destroy();
+      return new Uint8Array();
+    },
+  );
+
+  const {
+    issueCommandError,
+    vatCloseError,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  } = await issueCommandAndWait(worker);
+
+  t.not(issueCommandError, undefined, 'issueCommand() must produce and error');
+  t.not(vatCloseError, 'vat.close() must produce and error');
+  t.is(vatExitSignal, null, 'vat exit signal must be null');
+  t.is(vatExitCode, 2 /* E_IO_ERROR */, 'vat exit code must be 2');
+  t.is(
+    strErr,
+    'Got EOF on netstring read. Has parent died?\n',
+    'stderr must be "Got EOF on netstring read. Has parent died?"',
+  );
+});
+
+test('xsnap-worker complains while trying to WRITE when pipes are closed', async t => {
+  const worker = await spawnReflectiveWorker(
+    async function handleCommand(_message) {
+      // the worker is blocked on read here, so we should close "fromXsnap" pipe first
+      worker.fromXsnap.end();
+      worker.fromXsnap.destroy();
+      // write a fake but valid response into a pipe buffer,
+      // forcing the worker to move to the next state where it would
+      // attempt to write out return value from its own "handleCommand()"
+      // into a pipe which we've just closed above.
+      await makeNetstringWriter(makeNodeWriter(worker.toXsnap)).next([
+        QUERY_RESPONSE_BUF,
+        new Uint8Array(),
+      ]);
+      worker.toXsnap.end();
+      worker.toXsnap.destroy();
+      return new Uint8Array();
+    },
+  );
+
+  const {
+    issueCommandError,
+    vatCloseError,
+    strErr,
+    vatExitCode,
+    vatExitSignal,
+  } = await issueCommandAndWait(worker);
+
+  t.not(issueCommandError, undefined, 'issueCommand() must produce and error');
+  t.not(vatCloseError, 'vat.close() must produce and error');
+  t.is(vatExitSignal, null, 'vat exit signal must be null');
+  t.is(vatExitCode, 2 /* E_IO_ERROR */, 'vat exit code must be 2');
```

Comment:
```suggestion
  t.not(issueCommandError, undefined, 'issueCommand() must produce an error');
  t.not(vatCloseError, 'vat.close() must produce an error');
  t.is(vatExitSignal, null, 'vat exit signal must be null');
  t.is(vatExitCode, 2 /* E_IO_ERROR */, 'vat exit code must indicate an IO error');
```

Reply:
We might also abstract via `test.macro`:
```js
import { ExitCode } from '../api.js';
…
const testInterruption = test.macro(async (t, onRequest, expectedStderr) => {
  const worker = await spawnReflectiveWorker(msg => onRequest(worker, msg));

  const {
    issueCommandError,
    vatCloseError,
    strErr,
    vatExitCode,
    vatExitSignal,
  } = await issueCommandAndWait(worker);

  t.not(issueCommandError, undefined, 'issueCommand() must produce an error');
  t.not(vatCloseError, 'vat.close() must produce an error');
  t.is(vatExitSignal, null, 'vat exit signal must be null');
  t.is( 
    vatExitCode,
    ExitCode.E_IO_ERROR,
    'vat exit code must indicate an IO error',
  );  
  t.is(strErr, expectedStderr, 'stderr must indicate expected error');
});
      
test(
  'xsnap-worker complains while trying to READ an answer when pipes are closed',
  testInterruption,
  async function onRequest(worker, _message) {
    // the worker is blocked on read here, so we should close "fromXsnap" pipe first
    worker.fromXsnap.end();
    worker.fromXsnap.destroy();
    worker.toXsnap.end();
    worker.toXsnap.destroy();
    return new Uint8Array();
  },
  'Got EOF on netstring read. Has parent died?\n',
);
  
test(
  'xsnap-worker complains while trying to WRITE an answer when pipes are closed',
  testInterruption,
  async function onRequest(worker, _message) {
    // the worker is blocked on read here, so we should close "fromXsnap" pipe first
    worker.fromXsnap.end();
    worker.fromXsnap.destroy();
    // write a fake but valid response into a pipe buffer,
    // forcing the worker to move to the next state where it would
    // attempt to write out return value from its own "handleCommand()"
    // into a pipe which we've just closed above.
    await makeNetstringWriter(makeNodeWriter(worker.toXsnap)).next([
      QUERY_RESPONSE_BUF,
      new Uint8Array(),
    ]);
    worker.toXsnap.end();
    worker.toXsnap.destroy();
    return new Uint8Array();
  },
  'Caught SIGPIPE. Has parent died?\n',
);
```

Reply:
> We might also abstract via `test.macro`:

Cool!
I was thinking of something like this, but had no idea how to put it together so nicely.
Thank you!